src/main/java/ui/components/pickers/LabelPickerUILogic.java

```diff
@@ -60,58 +64,107 @@ private void populatePanes() {
     }
 
     public void toggleLabel(String name) {
-        addRemovePossibleLabel("");
+        removePossibleLabel();
         preProcessAndUpdateTopLabels(name);
         updateBottomLabels(""); // clears search query, removes faded-out overlay on bottom labels
         populatePanes();
     }
 
-    public void toggleSelectedLabel(String text) {
-        if (!bottomLabels.isEmpty() && !text.isEmpty() && hasHighlightedLabel()) {
+    public void toggleHighlightedLabel() {
+        if (!bottomLabels.isEmpty() && hasHighlightedLabel()) {
             toggleLabel(
                     bottomLabels.stream().filter(PickerLabel::isHighlighted).findFirst().get().getActualName());
         }
     }
 
-    @SuppressWarnings("PMD")
     public void processTextFieldChange(String text) {
-        String[] textArray = text.split(" ");
-        if (textArray.length > 0) {
-            String query = textArray[textArray.length - 1];
-            if (previousNumberOfActions != textArray.length || !query.equals(lastAction)) {
-                previousNumberOfActions = textArray.length;
-                lastAction = query;
-                boolean isBottomLabelsUpdated = false;
-
-                // group check
-                // TODO rewrite this to remove some nesting, and the PMD warning
-                if (TurboLabel.getDelimiter(query).isPresent()) {
-                    String delimiter = TurboLabel.getDelimiter(query).get();
-                    String[] queryArray = query.split(Pattern.quote(delimiter));
-                    if (queryArray.length == 1) {
-                        isBottomLabelsUpdated = true;
-                        updateBottomLabels(queryArray[0], "");
-                    } else if (queryArray.length == 2) {
-                        isBottomLabelsUpdated = true;
-                        updateBottomLabels(queryArray[0], queryArray[1]);
+        int curTextFieldLength = text.length();
+        if (curTextFieldLength > 0) {
```

Comment:
Consider putting the shorter branch first to improve understandability (i.e. pick low hanging fruits first). In this case, `if` is very long but `else` is very short. You can also use early returns to reduce nesting. e.g.
```
if (something){
   do this;
   return;
}